presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java

```diff
@@ -1461,50 +1521,147 @@ private static Object optimize(@Language("SQL") String expression)
+                return new VariableReferenceExpression(symbol.getName(), SYMBOL_TYPES.get(symbol));
+            }
+            return value;
         });
+
+        assertExpressionAndRowExpressionEquals(expressionResult, rowExpressionResult);
+        return expressionResult;
+    }
+
+    private static Object symbolConstant(Symbol symbol)
+    {
+        switch (symbol.getName().toLowerCase(ENGLISH)) {
+            case "bound_integer":
+                return 1234L;
+            case "bound_long":
+                return 1234L;
+            case "bound_string":
+                return utf8Slice("hello");
+            case "bound_double":
+                return 12.34;
+            case "bound_date":
+                return new LocalDate(2001, 8, 22).toDateMidnight(DateTimeZone.UTC).getMillis();
+            case "bound_time":
+                return new LocalTime(3, 4, 5, 321).toDateTime(new DateTime(0, DateTimeZone.UTC)).getMillis();
+            case "bound_timestamp":
+                return new DateTime(2001, 8, 22, 3, 4, 5, 321, DateTimeZone.UTC).getMillis();
+            case "bound_pattern":
+                return utf8Slice("%el%");
+            case "bound_timestamp_with_timezone":
+                return new SqlTimestampWithTimeZone(new DateTime(1970, 1, 1, 1, 0, 0, 999, DateTimeZone.UTC).getMillis(), getTimeZoneKey("Z"));
+            case "bound_varbinary":
+                return Slices.wrappedBuffer((byte) 0xab);
+            case "bound_decimal_short":
+                return 12345L;
+            case "bound_decimal_long":
+                return Decimals.encodeUnscaledValue(new BigInteger("12345678901234567890123"));
+        }
+        return null;
+    }
+
+    private static void assertExpressionAndRowExpressionEquals(Object expressionResult, Object rowExpressionResult)
+    {
+        if (rowExpressionResult instanceof RowExpression) {
+            // Cannot be completely evaluated into a constant; compare expressions
+            assertTrue(expressionResult instanceof Expression);
+
+            // It is tricky to check the equivalence of an expression and a row expression.
+            // We rely on the optimized translator to fill the gap.
+            RowExpression translated = TRANSLATOR.translateAndOptimize((Expression) expressionResult, SYMBOL_TYPES);
+            assertRowExpressionEvaluationEquals(translated, new ExpressionOptimizer(METADATA.getFunctionManager(), TEST_SESSION).optimize((RowExpression) rowExpressionResult));
+        }
+        else {
+            // We have constants; directly compare
+            assertRowExpressionEvaluationEquals(expressionResult, rowExpressionResult);
+        }
+    }
+
+    /**
+     * Assert the evaluation result of two row expressions equivalent
+     * no matter they are constants or remaining row expressions.
+     */
+    private static void assertRowExpressionEvaluationEquals(Object left, Object right)
+    {
+        if (right instanceof RowExpression) {
```

Comment:
This assumes the caller give the right order for left (expression) and right (row expression). It's better to add a check or switch order if the provided order is wrong.

Reply:
Actually both `left` and `right` are `RowExpression`. The one to compare `Expression` and `RowExpression` is `assertExpressionAndRowExpressionEquals`. It has a check of the order.